### ISSUE TITLE
Improve partial message support in json-file logging driver for > 16k log entries

### ISSUE THREAD

**Issue by bitsofinfo** [moby/moby#36777]
We have applications which frequently generate log messages > 16k in length and the json-file logging driver truncates these messages. (modsecurity audit logs, application stack traces etc) 

Docker daemon has a 16K buffer for log messages. If a message length exceeds 16K, it should be split by the json file logger and merged at the endpoint.

For reference can functionality similar to this be implemented for the `json-file` logging driver? Some option that would add new fields that allows a consumer of the logs to re-correlate split/partial log messages that are > 16k in length overall. 

https://github.com/moby/moby/pull/35831
https://github.com/elastic/beats/issues/6605
https://github.com/moby/moby/pull/34888#issuecomment-378346970

**Comment by ruflin** [comment]
From Filebeat perspective it would be simplest if the `json-file` driver would support the `SizedLogger ` interface: https://github.com/moby/moby/pull/34888/files#diff-3257fea85e404eee106b785fa9d14f68R86 This means the buffer size could be configured. Then Filebeat would not have to try figuring out which events belong together.

**Comment by nfedyk** [comment]
I've experienced the same issue picking up logs for mod security, as they often exceed 16k. Can we get this fixed?

**Comment by ruflin** [comment]
@cpuguy83 The way I understood the `SizedLogger` interface is that each logger could implement this to overwrite the bufSize. So far the only logger that does this is the `awslogs` driver. I assumed we could add a `BufSize() int` method to the `json-file` logger which would read the buffer size from an optional flag. All the defaults would stay the same.

I understand the concern about bigger and bigger log buffer but the interface is already there and in use by `awslogs`. If it is configurable the requests for a bigger buffer size should disappear. The side effect would be that we potentially make it possible for a user to shoot in his own foot with a very large buffer size, but I would assume the users that will change this value hopefully understand what the side effects can be. Is the concern more related to the `json-file` logger? It seems the `json-file` logger is kind of special in docker and I definitively lack the bigger picture here.

**Comment by cpuguy83** [comment]
@ruflin 

> but the interface is already there

The interface was added so that log drivers that already buffer their own messages (for various reasons) can tune the log copier so that the buffer aligns with the driver.

The json log driver only needs to be enhanced to better support partial log messages. It's unfortunate that the log presentation format on the API might be rather difficult to bolt on such support for so (docker API) clients can deal with partials more easily as well.

**Comment by bmudda** [comment]
why not do both, what is the real harm in having an option for the end user to tweak the buffer size, in addition to partial options?

**Comment by cpuguy83** [comment]
Lots of oversized log buffers consuming all available memory, which is why we have a capped buffer to begin with.

If partial messages are difficult to parse, then we need to fix that. Increasing the buffer size to reduce the possibility of a partial is a hack that defeats the purpose.

**Comment by ruflin** [comment]
@cpuguy83 Thanks for providing some more background here. Reading into more details in https://github.com/moby/moby/pull/34913 and https://github.com/moby/moby/pull/34888 I now better understand the intention that it was designed for the logger to set the buffer size. 

For the partial logging let me explain why it's tricky from a filebeat perspective. So far we have 2 types of log messages: single line logs with a new line at the end or multiline which are defined by a pattern. For the single line logs, each line is read and only after the reading phase the json is potentially decoded. The line reading already continued during this time and is decoupled for the decoding. Reading line by line will not work for the above as we decode the json too late.

With multiline we could probably create a pattern that matches the partial flag and get both lines into one message. But normal multlines are then actually one message and don't required logic to be put together. I think we could even argue docker multiline is perhaps superior to other multiline as it's structured data but it does not work with any of the existing methods we have in place. What would make it work with our existing method is if we would break the json across 2 lines but that seems to be a pretty bad option :-( I need to think more about what we could do if we split message to make it more consumer friendly.

Alternative is to get our own logger in place and hook into Docker directly instead of the files. @thaJeztah You commented on the other issue that the files created by the json-file logger should not be consumed by other tools. Why is that?

**Comment by cpuguy83** [comment]
@ruflin Everything in /var/lib/docker is considered internal and could change at any time. It was never intended for external processes to consume these files.
While we probably won't be messing with json-file stuff (lest we break a bunch of K8s users at the very least), it's definitely not a good idea to poke around in there.

**Comment by bitsofinfo** [comment]
> While we probably won't be messing with json-file stuff 

So are you saying that adding the partial flags is not even possible?

**Comment by cpuguy83** [comment]
Adding new things is fine.

**Comment by towolf** [comment]
64k please, I will deal with the memory consumption.

The argument, that we'll have to bump and bump again is not so convincing.

**Comment by bitsofinfo** [comment]
Agreed. Not seeing any harm in simply adding an option, providing the end *user* a choice to adjust the buffer size for their own needs. If they shoot themselves in the foot, their problem. The implications of mis-configuring this via the option could just be clearly stated in the documentation w/ adequate warnings.

**Comment by thaJeztah** [comment]
For elasticsearch; also have a look at https://github.com/moby/moby/issues/25694#issuecomment-354138086 (dedicated logging driver plugin for elasticsearch)

**Comment by cpuguy83** [comment]
Yes, unless you actually need to read the real log files for some reason
(which should be a no-no, but has been done for years mostly due to Kube)
then it's best to use a log driver that can deal with this.
We can add extra metadata about partial messages in the json if that will
help.

Additionally, json log messages which do not have a `\n` should be
considered partial. I'm not sure what/how you might need to figure in
filebeat to support this (likewise for how to deal with extra partial
metadata in the json), but that is what can be done.

**Comment by optical** [comment]
Thanks @thaJeztah. We prefer to log to disk before shipping the logs over to elasticsearch, as it can help prevent loss of important error logs during outages, and outages tend to be the times you really need your logs. 

Thanks @cpuguy83, I didn't actually notice the trailing \n at the end of the message. It looks like this is indeed feasible then. My issue looks like it belongs with filebeat rather then docker in this case.

**Comment by towolf** [comment]
@optical you may be happy to hear that a [PR is already in progress](https://github.com/elastic/beats/pull/6967) for Filebeat to re-join split partial messages.

**Comment by optical** [comment]
Looks good @towolf, thanks!

**Comment by ruflin** [comment]
The key part I also initially missed is that `\n` can be used for the differentiation of partial lines. We will try to get the change into Filebeat rather soon and then are looking for some testers ;-)

**Comment by dsx0** [comment]
i commented my solution at: https://github.com/moby/moby/issues/34620#issuecomment-619369707
hope it helps.
